third-party/pvm/pvm-3.4.6/pvm3/man/man1/pvm_shmd.1
.\" $Id: pvm_shmd.1,v 1.2 2004/02/17 17:53:05 pvmsrc Exp $
.TH PVM_SHMD 1PVM "11 March, 1999" "G.E.Fagg" "PVM Version 3.4"
.SH NAME
pvm_shmd \- PVM shared memory (2) daemon
.SH SYNOPSIS
.ft B
pvm_shmd
[
\-options
]
[
hostfile
]
.ft R
.SH DESCRIPTION
.I Pvm_shmd
is a daemon process which maintains and handles the usage of shared memory
resources (memory segments, semaphores, message queues) on a particular PVM
host.
One \fIpvm_shmd\fR must run on each host in a virtual machine that wants to 
use the shmd shared memory message passing layer. The \fIshmd\fR message
passing layer allows local processes to a host to use shared memory for
message passing on that host ONLY. See the shmd directory README for further
notes and restrictions.

The daemon can be started manually as with any other PVM task. 
Only one \fIpvm_shmd\fR can be active per host. If the \fIpvm_shmd\fR
detects another \fIpvm_shmd\fR it will shutdown.

The \fIpvm_shmd\fR can be killed from the PVM console by using the reset
command.
Local and remote pvm_shmds can also be started from the PVM console
program \fIpvm\fR.

The name of the daemon executable is \fIpvm_shmd\fR.

.SH Shared Resource Usage
The \fIpvm_shmd\fR will attempt to create upto \fIMAXSEG\fR shared memory
segments of upto \fIMAXPAGES\fR memory pages in size. The size of each memory
page is architecture dependent(see getpagesize(3C)). 
The segments are then numbered 0..N-1 where N is the number of segments created.
The first segement (0) has the shared memory control structure placed in its
head. This structure allows alien processes to located the other shared memory
segments and any required controlling information.

The message passing layer, allocates memory from these segments asynchronously
without any interaction with the \fIpvm_shmd\fR using semaphores to protect
data during updates to any associated structures. 
For each segment there is an associated page map of which processes have
currently locked a page. Each page map for a segment has a separate semaphore
protecting it. The semaphores are accessed with the SEM_UNDO flag set so that
if a process holding a semaphore should die, the OS (should) reset the
semaphore automatically, thus allowing any waiting/blocked processes to
continue.

The \fIpvm_shmd\fR only
maintains the segments and their allocation page map(s).
Thus if a process allocated pages in a segment and then exits, it is the 
\fIpvm_shmd\fR that detects this and then frees the allocated pages.

The \fIpvm_shmd\fR can have its status checked at any time by using the 
\fIpvm_shmd_stat\fR process to kick it into reporting onto either stdio or the
pvm log file (pvml.uid) its internal state.

The shared memory and all associated processes can be cleared by sending the 
\fIpvm_shmd\fR a HUP signal.

Sending the \fIpvm_shmd\fR a TERM signal will just cause it to clear any shared
resources and then exit.

The \fIpvm_shmd\fR cannot catch the KILL signal.

If the \fIpvm_shmd\fR is killed without clearing all of its shared resources
these can be cleared by calling \fIipcfree\fR which resides in the pvm3/lib
directory.
.PP

.SH Message Passing using SHMD
The \fIpvm_shmd\fR handles resources that are used by special versions of 
pvm_psend() and pvm_precv() stored in the libpvmshmd.a library. Thus to use
these facilities, applications have to link to this library instead of the
usual libpvm3.a library.
.PP

.SH OPTIONS
The following options may be specified on the command line when
starting the pvm_shmd:
.TP 8
-debug=\fIlevel\fR
Sets the \fIpvm_shmd\fR debug level.
Used to debug the pvm_shmd or libpvmshmd (not intended to be used to debug
application programs).
.PP
.TP 8
-maxsegs=\fImaxsegs\fR
Sets the maximum number of segments that the \fIpvm_shmd\fR can create. This is
used to over-ride the compiled in value from shmd.h. Note that the value cannot
be above the MAXSEGS in the shmd.h file.
.TP 8
-maxpages=\fImaxpages\fR
Sets the maximum segment size to \fImaxpages\fR pages of memory. This value
cannot be above the compiled value MAXPAGES in shmd.h or the actual OS defined
limit.

.SH NOTES
Remember that \fIpvm_shmd\fR allocated memory from the VM available on the
machine. Allocating more segements improves performance as there is less
sharing of segments (semaphores for their page maps). Although you must
remember to leave some memory available for normal program and OS system
usage, as the \fIpvm_shmd\fR allocated memory is *ONLY* used for message
passing.
.SH FILES
.ta 4 40
.nf
  $PVM_ROOT/lib/$PVM_ARCH/pvm_shmd PVM shared memory daemon executable
  $PVM_ROOT/shmd/shmd.h Shared Memory hard limits header file
  /tmp/pvml.\fIuid\fR Pvmd runtime error log
.fi

.SH SEE ALSO
pvm(1PVM),
pvmd(1PVM),
getpagesize(3C),
ipcs(1),
msgctl(2),
semctl(2),
shmctl(2),
signal(5)



CHANGES
====================
Release Changes List
====================

===========
version 0.9
===========

Second public release of Chapel compiler, April 16, 2009

High-Level Themes:
------------------
- multi-locale task parallelism
- very initial/limited support for data parallelism, distributed arrays
- improved stability, portability, and performance
- target audience: general public

Environment Changes:
--------------------
- added a script to print out set and inferred CHPL env vars
  (see util/printchplenv.pl)
- added support for the IBM xlc/xlC compilers by setting CHPL_*_COMPILER to ibm
- added an environment variable for naming a queue for the PBS launcher
  (see doc/platforms/README.xt-cle)
- added an environment variable for specifying #cores/locale to pbs/aprun
  (see doc/platforms/README.xt-cle)
- improved the logic for inferring host and target compilers for xt-cle

Changes to Chapel Language:
---------------------------
- added a typeToString() function for converting a static type into a string

Newly Implemented Features:
---------------------------
- added queries for task/thread counts per locale for CHPL_THREADS=pthreads
  (see doc/README.threads)
- added a printMemTable() routine to the Memory module for use in Chapel
  (see doc/README.executing)
- added a prototype multidimensional Block distribution: MultiBlockDist.chpl
  (see examples/tutorials/block2D.chpl)

Portability of code base:
-------------------------
- improved our general method of building and using GASNet for portability
- improved portability of GASNet configure + build steps for xt-cle
- improved the heuristic for setting GASNET_MAX_SEGSIZE for xt-cle
  (see doc/platforms/README.xt-cle)
- added initial support for IBM Power5 and Power6 systems
  (see doc/platforms/README.ibm)
- made flex and bison generate .cpp files rather than .c files for portability
- added capabilities to build Chapel on Macs using xcodebuild
  (see doc/platforms/README.macosx)
- added a -d option to our use of aprun in the aprun and pbs launchers
- removed the assumption that function pointers are identical between nodes

Syntactic/Naming Changes:
-------------------------
- disallowed identifiers that start with $
- removed compilerError and compilerWarning as keywords
- permitted binary and hex literals to be written using uppercase B/X
- renamed classes used for defining distributions

Runtime Library Changes:
------------------------
- added support for communication diagnostics and debugging
  (see doc/technotes/README.comm-diagnostics)

Documentation:
--------------
- added a new ACKNOWLEDGEMENTS file for financial and machine time support
- refactored the doc directory, pushing platform & technical notes into subdirs
- minor changes and improvements to the language specification
- minor edits to many of the README files
- updated the man page

Example Codes:
--------------
- pushed most top-level examples into tutorials/ and programs/ subdirectories
- added new tutorial examples for variables, ranges, classes, generic classes
- added new tutorial examples for timers and the prototype Block distribution
- changed quicksort to ensure levels beyond thresh do not re-enable parallelism

Compiler Flags:
---------------
- added a flag to disable/enable the remote value forwarding optimization

Bug Fixes/New Semantic Checks (for old semantics):
--------------------------------------------------
- fixed a number of spurious errors caused by how we were using GASNet
- fixed problems in how we terminated multithreaded GASNet programs
- added the ability for _chpl_comm_fork[_nb] to handle large arguments
- applied a GASNet patch that fixes an x86 alignment error
- fixed a bug which had previously limited execution to 1024 locales
- fixed a bug in which arrays could only store 2**31 elements
- eliminated a race condition when two tasks initialize the same module
- fixed a bug in which integer exponentiation was being computed using pow()
- fixed a bug in which slicing via .. could only be done for int(32) indices
- fixed a bug with Block1D domains that exceed the bounding box argument
- added a check to ensure that return statements are not used at module scope
- improved verification of formal/actual argument types
- fixed a bug in which long program names broke the PBS launcher
- fixed a bug with incorrectly spacing operators in generated code
- fixed Chapel error messages that referred to C's errno variable
- fixed a bug with accidentally shared state during parallel iterations
- fixed a bug related to reductions involving forall expressions

Error Message Improvements:
---------------------------
- added a runtime error message if the launcher binary can't find the _real one

Compiler Analysis and Optimizations/Performance of Generated Code:
------------------------------------------------------------------
- improved the correctness and effectiveness of remote value forwarding
- optimized performance of local communications
- optimized parallel iteration for Block1D to minimize remote references
- parallelized declarations of Block1D distributions, domains, and arrays
- optimized implementation of broadcasts used to initialize private data
- optimized away a thread per locale used for progress+termination
- improved reclamation of compiler-allocated memory
- improved the correctness and effectiveness of re-privatization optimizations
- privatized the default distribution object
- removed array/domain wrapper records

Compiler Performance and Cleanup:
---------------------------------
- removed hand-baked and occasionally incorrect integer types int32, int64
- fixed an infinite loop when inlining recursive iterators
- fixed cases in which compiler temps were used outside the variable's scope

Cleanup of Generated Code:
--------------------------
- removed redundant reference types from the generated code

Internal:
---------
- switched iterator implementation to use records rather than classes

third-party/pvm/pvm-3.4.6/pvm3/shmd/pvm_shmd.c

static char rcsid[] =
	"$Id: pvm_shmd.c,v 1.2 1999/03/12 20:22:39 pvmsrc Exp $";

#include <stdio.h>
#include <sys/types.h>
#include <sys/ipc.h>
#include <sys/shm.h>
#include <sys/msg.h>
#include <sys/param.h>
#include <errno.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <signal.h>

#include <pvm3.h>

/* GEF libs */
#include "shmd.h"

/* Prototypes */
unsigned int	get_max_seg_size();
unsigned int	create_segments();
void		clean_segs();
int		make_mqueue();
int		del_mqueue();
void 		bambismother();
void 		huphuphup();

/* Consts */
/* #define MAXSEGS 255		*/ 
/* in the header files now */
/* well its a limit and we do need one */
						/* can be controlled by ifdef or ENV vars */
						/* at run-time */

#ifndef SHM_FAILED
#define SHM_FAILED (void*)-1
#endif

#define FAILEDCOUNT (10)

/* Globals */
unsigned segsize;
unsigned numsegs;
key_t keys[MAXSEGS];
int shmids[MAXSEGS];
void* segptrs[MAXSEGS];
unsigned totalmem;



main(argc, argv) 
int argc;
char *argv[];
{
	int i,id;
	key_t	key;
	int		semid;
	struct shmid_ds shmsbuf;
	unsigned size;
	int pagesize;
	size_t bsize, bsizepb, o1, o2, o3;		/* for working out mem offsets */
	int pages[MAXSEGS];
	char* addr;
	unsigned  firstpagehdr;
	int tid;
	int bufid, from, bytes, msgtag;
	int died;
	int tidslots; /* not an array, but number of tids I can handle at once */
	int slot, left, join;
	key_t mqkey;
	int  mqid;
	int j, cc;
	int finished=0;

	/* first things first */

	/* The PVM stuff */
	tid = pvm_mytid();
	if (tid<=0) { 
		fprintf(stderr,"Cannot join the virtual machine.\n");
		exit (-3);
	}	

	if (register_shmd (tid)) {
		fprintf(stderr,"Cannot setup system mailbox entry.\n");
		exit (-4);
	}	




	o1 = (size_t) (16*1024); /* first offset is a 16K block */

	/* calculate offsets into shmpages */
	pagesize = getpagesize();
	bsize = sizeof (blockinfo_t);
	/* calculate blockinfo size in whole pages (on page boundary) */
	if (bsize%pagesize) bsizepb = ((bsize/pagesize)+1) * pagesize;
	else bsizepb = (bsize);
		
	key = (key_t)	 (0x2d000000+((int)getuid()*0x1000));	/* inital key */
	/* segsize = get_max_seg_size(); */
	segsize = get_max_seg_size(pagesize);
/* printf("seg size %d %x\n", segsize, segsize);
exit (-1); */
	/* numsegs = create_segments(key, keys, shmids, segptrs, MAXSEGS); */
	numsegs = create_segments(key, MAXSEGS);
	if (numsegs<2) {
		fprintf(stderr,"WARNING:\nCan not create 2 or more segments.\nExiting.\n");
		fflush(stderr);
		clean_segs();
		exit (-2);
	}

	for(i=0;i<numsegs;i++) 
		printf("Base addresses are [0x%x]\n", segptrs[i]);

	/* first segment is not all available for messages as it has a header */
	firstpagehdr = o1+(bsizepb*numsegs);
	semid = init_sem(0x4000, 1);	/* todo */
	addr = (char*)segptrs[0]+(o1);
	printf("Blockinfo addresses are [0x%x]\n", addr);
	pages[0] = initblocks ( addr, pagesize, (segsize-firstpagehdr)/pagesize, 
				0x4000+0, keys[0], semid, shmids[0]);

	((blockinfo_t*)addr)->segoffset = firstpagehdr;	
				/* first page starts after header */

	/* First page offsets */
	((blockinfo_t*)addr)->nextoffset = bsizepb;
		/* and the previous is wrapped around... */
	((blockinfo_t*)addr)->previousoffset = (numsegs-1) * bsizepb; 

	/* all other segments are just for message info */
	for(i=1;i<numsegs;i++) {
		semid = init_sem(0x4000+i, 1); /* todo */
		addr = (char*)segptrs[0]+(o1)+(bsizepb *i); 
		printf("Blockinfo addresses are [0x%x]\n", addr);
		pages[i] = initblocks ( addr, 
			pagesize, (segsize/pagesize), 0x4000+i, keys[i], semid, shmids[i]);
		((blockinfo_t*)addr)->segoffset = 0;	
			/* i.e. first page is at the head of the seg */

		/* offsets to previous / next blockinfo structs */
		if (i<(numsegs-1)) 
			((blockinfo_t*)addr)->nextoffset = bsizepb;
		else 	/* wrap around */
			((blockinfo_t*)addr)->nextoffset = -((numsegs-1) * bsizepb);

		((blockinfo_t*)addr)->previousoffset = -bsizepb;
	}	

	/* previous/next wrap around tests */
	/* first forwards */
	printf("Forwards loop\n");
	addr = o1 + (char*)segptrs[0];
	for(i=0;i<=numsegs;i++) {
		printf("segkey [0x%x]\t", ((blockinfo_t*)addr)->segkey );
		addr += ((blockinfo_t*)addr)->nextoffset;
	}
	printf("\n\n");

	/* backwards */
	printf("Backwards loop\n");
	addr = o1 + (char*)segptrs[0];
	for(i=0;i<=numsegs;i++) {
		printf("segkey [0x%x]\t", ((blockinfo_t*)addr)->segkey );
		addr += ((blockinfo_t*)addr)->previousoffset;
	}
	printf("\n\n");




	for(i=0;i<numsegs;i++) {
		addr = (char*)segptrs[0]+(o1)+(bsizepb *i); 
		if (pages[i]) blockmap (addr);
	}
		
/* 	totalmem = segsize * numsegs; totalmem /= (1024*1024);  */
	/* MB */

	printf("Pages available in each segment: ");
	totalmem = 0;
	for(i=0;i<numsegs;i++) {
		printf("seg[%d]:[%d]\t", i, pages[i]);
		totalmem += pages[i];
	}
	totalmem = ((totalmem * pagesize) / 1024);

	printf("Seg size in bytes [%d] and number of segments [%d]\nTotal message buffer memory [%d KBytes]\n",
			segsize, numsegs, totalmem);


	/* Now we can fill in that first section of memory */
	tidslots = init_tidinfo ((char *)segptrs[0], 
							(16*1024));  /* make tidinfo struct */

	/* missed out a few details.. */
	((shm_header_info_t *) segptrs[0])->numsegs = numsegs;

	/* should check tidslots here TODO */

	/* Safe to register myself now */

	if (register_shmd (tid, shmids[0])) {
		fprintf(stderr,"Cannot setup system mailbox entry.\n");
		exit (-4);
	}	

	/* now the pvm daemon knows we are here */
	/* when it exits it sends up a TERM so that we can exit neatly */
	/* we also catch a HUP for completeness */ 
	(void) signal (SIGTERM, bambismother);
#ifndef LINUX
	(void) sigset (SIGTERM, bambismother); /* yes I mean it, ok ? */
#endif
	(void) signal (SIGHUP, huphuphup);

	while (!finished) {
		/* ok, we really would do some work in here... honest */
		bufid = pvm_recv(-1,-1);
		pvm_bufinfo( bufid, &bytes, &msgtag, &from );
		switch (msgtag) {
			case 1: /* join */
				join = from;
				make_mqueue (join, &mqkey, &mqid);
				if (mqid<0) {
					pvm_initsend(PvmDataDefault);
					pvm_pkint (&mqid, 1, 1);  /* failed */
					pvm_send (join, join);
					fprintf(stderr,"TID [0x%x] failed on make_mqueue\n",
						join);
					break;
				}

				slot = add_tidinfo (join, mqkey, mqid);

				if (slot<0) {
					pvm_initsend(PvmDataDefault);
					pvm_pkint (&slot, 1, 1);  /* failed */
					pvm_send (join, join);
					fprintf(stderr,"TID [0x%x] failed on add_tidinfo\n",
						join);
					break;
				}

				/* else ok, so we setup notify and then signal task */

				printf("TID [0x%x] added into tidinfo slot [%d]\n", join, slot);
				pvm_notify (PvmTaskExit, 9, 1, &join);
				fprintf(stderr,"Adding [0x%x] to notify list\n", join);
				fflush(stderr);

				j = 0;	/* i.e. isitok is yes */
				pvm_initsend(PvmDataDefault);
				pvm_pkint (&j, 1, 1);  /* passed */
				fprintf(stderr,"Sending message [%d] to [0x%x] tag [0x%x]\n",
						j, join, join);
				cc = pvm_send (join, join);
				fprintf(stderr,"Sent message [%d] to [0x%x] tag [0x%x] = [%d]\n",
						j, join, join, cc);
				fflush(stderr);
				break;

			case 2: /* exit (polite) */

				left = from;	/* just adds to readability, maybe  */
				del_mqueue (left);
				slot = remove_tidinfo (left);
				printf("TID [0x%x] removed from tidinfo slot [%d]\n", left, slot);

				/* freeing space by this task */
				left = left & 0xFFFF; /* task part of address */
				for(i=0;i<numsegs;i++) {
					addr = (char*)segptrs[0]+(o1)+(bsizepb *i); 
					if (pages[i]) freeblock_by_id (addr, left);
				}

				((shm_header_info_t *)segptrs[0])->tidsexited++;
				break;

			case 3: /* blockmap */
				for(i=0;i<numsegs;i++) {
					addr = (char*)segptrs[0]+(o1)+(bsizepb *i); 
					if (pages[i]) blockmap (addr);
				}
				break;
		    case 9: /* TERM and caught by pvmd with notify */
				/* Find out who died */
				pvm_upkint(&died, 1, 1);
				
				del_mqueue (died);
				slot = remove_tidinfo (died);
				printf("TID [0x%x] removed from tidinfo slot [%d]\n", died, slot);

				printf("Task [0x%x] exited. Deleting any blocks allocated.\n",
						died);

				/* freeing space by this task */
				died = died & 0xFFFF; /* task part of address */

				for(i=0;i<numsegs;i++) {
					addr = (char*)segptrs[0]+(o1)+(bsizepb *i); 
					if (pages[i]) freeblock_by_id (addr, died);
				}
				
				((shm_header_info_t *)segptrs[0])->tidsmurdered++;

				break;


			default: /* ? */
				fprintf(stderr,"Unknown message from [0x%x] of tag [0x%x]\n",
						from, msgtag);
				pvm_freebuf (bufid);
				break;

			} /* end switch */

	}

	clean_segs();
	pvm_exit(); /* which takes the PVM reg with it! */
	exit (0);	/* Normally I hope */
}




/* unsigned int	get_max_seg_size() */
unsigned int	get_max_seg_size(psize)
unsigned psize;
{
	key_t key;
	unsigned int	size;
	int	id;

	key = (key_t)	 0x2d000000;	
/* 	size = 1024 * 1024 * 1024; */
/*	size = 1024 * 1024 * 32;		*/
/* start looking from 32MBytes */
	size = MAXPAGESALLOWED * psize;	/* start from our limit */

	for(;;) {
		id = shmget (key, size, IPC_CREAT | 0600);

		if (id>0) {
/* 		printf("id\t%d\n",id); */
/* 		printf("key\t%x\n",key); */
/* 		printf("size\t[0x%x or %u bytes] [%u KBytes]\n", size, size, size/1024); */
		break;
		}
/* 	key += 0x10; */

	   if (id<0) {
		   if (errno==EINVAL) perror("shmget:EINVAL");
		   if (errno==ENOSPC) perror("shmget:ENOSPC");
		   size /= 0x2;	
/* 		   break; */
		  }
	}
	shmctl (id, IPC_RMID, 0);	/* I know its a waste.. but.. */

	return (size);
}

unsigned int create_segments(basekey, maxsegs)
key_t	basekey;
unsigned int maxsegs;
{
	int c=0;	/* number of segments that we get a valid id for */
	int v=0;	/* number of segments that we get a valid ptr for */
	void *ptr;
	key_t key;
	struct shmid_ds shmsbuf;
	int suc, attached;
	int i;
	int f;	/* f is for failed */
	int num;	/* number of segments that we have a valid id and valid ptr for */


	key = (key_t) basekey;

	/* make sure we are all reset fully */
	for(i=0;i<maxsegs;i++) { keys[i] =-1; shmids[i] = -1; segptrs[i] = (void*)0; };

	/* loop until we get a valid key for each segment */
	f = 0;
	for(c=0;c<maxsegs;) {
		keys[c] = basekey; 	/* this one gets this key */
		basekey += 0x10;	/* the next one gets a different one */
	
			
		shmids[c] = shmget (keys[c], segsize, IPC_CREAT | 0600);
		if ((int)shmids[c]==-1) { 
			perror("shmget");
			f++;
		}
		else {	/* we got an id ok */
			f = 0;	/* reset failed count */
			c++;	/* inc key count */
		}
		if (f==FAILEDCOUNT) {
			fprintf(stderr,"[create_segments] failed shmget() count reached.\n");
			break;
		}

	}	
			
		
	/* Now to try and attach to these segments */		
	for (v=0;v<c;v++) {
/* 		segptrs[v] = shmat (shmids[v],0,SHM_RDONLY); */
		segptrs[v] = shmat (shmids[v],0,0);
/* 		if ((int)segptrs[v]==-1) {  */
		if (segptrs[v]==SHM_FAILED) { 
			perror("shmat");
			fprintf(stderr,"shmat error code [%d]\n", segptrs[v]);
			segptrs[v] = (void *) 0;
		}
	}	

	
/* sleep(1); */
#ifdef DEBUG
	for (i=0;i<c;i++) {
		printf("id\t%d\n",shmids[i]);
		printf("key\t%x\n",keys[i]);
		printf("ptr\t0x%x\n", segptrs[i]);
	}
#endif

	/* now to get rid of the ones that don't count! */
	num = 0; /* meaning none */
	for(i=0;i<(c);i++)
		if (!segptrs[i]) {
			shmctl (shmids[i], IPC_STAT, &shmsbuf);
			attached=shmsbuf.shm_nattch; 
			suc = shmctl (shmids[i], IPC_RMID, 0);
			shmids[i] = -1;
			keys[i] = -1;
		}
		else /* i.e. its ok */
			num = i+1;	/* i+1 as i is from 0..N-1 not 1..N ! */


	return (num);

}

void clean_segs () 
{
int i;

	for (i=0;i<MAXSEGS;i++) shmctl (shmids[i], IPC_RMID, 0);
	/* gone */
}
	

int	register_shmd (tid, shmid)
int tid;
int shmid;	/* id of first shm seg that has all syshrd info in it */
{
    int cc;
    int sbuf,ssbuf;
	char fullname[1024];
	char hname[MAXHOSTNAMELEN];


	bzero (fullname, 1024);
	gethostname(hname, MAXHOSTNAMELEN);
	strcat (fullname, "pvm_shmd:");
	strcat (fullname, hname);

	printf("Registering process [0x%x] as [%s]\n", tid, fullname);
	if (shmid>0) 
		printf("%s storing syshdr as shmid [0x%x]\n", fullname, shmid);
	else {
		printf("%s marking shmem as disabled\n", fullname, shmid);
		 shmid = -1;
		 }


    ssbuf = pvm_setsbuf(0);
    sbuf = pvm_initsend(PvmDataDefault);
    pvm_pkint(&tid,1,1);
    pvm_pkint(&shmid,1,1);

    if ( (cc = pvm_putinfo(fullname, sbuf, PvmMboxDefault)) != 0 )
    {
        pvm_perror("pvm_shmd already running?");
    }

    pvm_setsbuf(ssbuf);

    pvm_freebuf(sbuf);

    return(cc); /* Will be PvmOK if register succeeded, < 0 otherwise */
}

int	make_mqueue (tid, mqkey, mqid) 
int tid;
key_t* mqkey;
int* mqid;
{
key_t key;
int  id;
int i,j,k;

	key = (key_t)    (0x2c000000+((int)getuid()*0x1000)+(0xFFFF&tid));   
			/* inital key */

	for(i=0;i<1000;i++) {
		id = msgget(key, IPC_CREAT|0666);
		if (id>0) break;
		else
			key++;
	}

	if (id<0) return (-1);	/* can't get one ! */

	/* else we can and we did! */
	printf("TID [0x%x] has a new message queue of key [0x%x] and id [%d]\n",
			tid, key, id);

	*mqkey = key;
	*mqid  = id;

	return (0);
}


int	del_mqueue (tid) 
int tid;
{
key_t key;
int  id;
int i,j,k;

	/* first get the id from the tidinfo structure using the processes tid */
	
	id = find_tidinfo_id (((char *)segptrs[0]), tid);

	if (id<0) {
		fprintf (stderr, "No TID [0x%x] found when removing mqueue ??\n",
				tid);
		fflush(stderr);
		return (-1); /* i.e. no tid record... ? */
	}
	
	i = msgctl (id, IPC_RMID, (struct msqid_ds *) 0 );

	if (!i) {
		printf("TID [0x%x] had a message queue of id [%d] deleted.\n",
				tid, id);
		return (0);
	}
	else {
		printf("Failed to remove TID [0x%x] message queue [%d]\n",
			tid, id);
		return (-1);
	}

}

void huphuphup ()
{
fprintf(stderr,"pvm_shmd: Received a sig HUP.\nExiting.\n");
fflush(stderr);
clean_segs();
pvmendtask();
exit (SIGHUP); 
}

void bambismother ()
{
fprintf(stderr,"pvm_shmd: Received a sig TERM.\nExiting.\n");
fflush(stderr);
clean_segs();
pvmendtask();
/* pvm_exit(); */
/* this causes a PVM not implemented error to appear */
exit (SIGTERM); 
}

